docs/source/distributed.pipelining.rst

```diff
@@ -4,182 +4,310 @@
+*********************
+
+In the `PiPPy <https://github.com/pytorch/PiPPy>`_ repo where this package was
+original created, we kept examples based on unmodified Hugging Face models.
+See the `examples/huggingface
+<https://github.com/pytorch/PiPPy/tree/main/examples/huggingface>`_ directory.
+
+Examples include:
+
+* `GPT2 <https://github.com/pytorch/PiPPy/tree/main/examples/huggingface/pippy_gpt2.py>`_
+* `Llama <https://github.com/pytorch/PiPPy/tree/main/examples/llama>`_
+
+
+Technical Deep Dive
+*******************
+
+How does the ``pipeline`` API split a model?
+============================================
+
+First, the ``pipeline`` API turns our model into a directed acyclic graph (DAG)
+by tracing the model.  It traces the model using ``torch.export`` -- a PyTorch 2
+full-graph capturing tool.
+
+Then, it groups together the **operations** and **parameters** needed by a stage
+into a reconstructed submodule: ``submod_0``, ``submod_1``, ...
+
+Different from conventional submodule access methods like ``Module.children()``,
+the ``pipeline`` API does not only cut the module structure of your model, but
+also the **``forward``** function of your model.
+
+This is necessary because model structure like ``Module.children()`` merely
+captures information during ``Module.__init__()``, and does not capture any
+information about ``Module.forward()``. Said differently, ``Module.children()``
+lacks information on the following aspects which are key to pipelininig:
+
+* Exectuion order of child modules in ``forward``
+* Activation flows between child modules
+* Whether there are any functional operators between child modules (for example,
+  ``relu`` or ``add`` operations would be be captured by ``Module.children()``).
+
+The ``pipeline`` API, on the contrary, makes sure that the ``forward`` behavior
+is truly preserved. It also captures the activation flow between the partitions,
+helping the distributed runtime to make correct send/receive calls without human
+intervention.
+
+Another flexibility of the ``pipeline`` API is that split points can be at
+arbitrary hierarchy of your model. In the split partitions, the original model
+hierarchy related to that partition will be reconstructed at no cost of yours.
+At a result, fully-qualified names (FQNs) pointing to a submodule or parameter
+would be still valid, and services that relies on FQNs (such as FSDP, TP or
+checkpointing) can still run with your partitioned modules at almost zero code
+change.
+
+
+
+Implementing Your Own Schedule
+******************************
+
+You can implement your own pipeline schedule by extending one of the following two class:
+
+* ``PipelineScheduleSingle``
+* ``PipelineScheduleMulti``
+
+``PipelineScheduleSingle`` is for schedules that assigns *only one* stage per rank.
+``PipelineScheduleMulti`` is for schedules that assigns multiple stages per rank.
+
+For example, ``ScheduleGPipe`` and ``Schedule1F1B`` are subclasses of ``PipelineScheduleSingle``.
```

Comment:
I think we need to say more about what it takes to write a schedule. 

- the API you have to implement.  (step_batch, or something?) and how 'step' API will automatically work on top
- the expectations of the 'step' API- mainly that it calls the particular stage APIs for forward/backward in a valid sequence including getting their comm ops
- some pointer on using batch_p2p correctly.  (maybe including a schedule picture with crosses and how that is useful to design batching)